libraries/botbuilder/src/cloudAdapter.ts

```diff
@@ -0,0 +1,351 @@
+     * @param callerId Optional, the caller ID
+     * @param retryCount Optional, the number of times to retry a failed connection (defaults to 7)
+     */
+    async connectNamedPipe(
+        pipeName: string,
+        logic: BotLogic,
+        appId: string,
+        audience: string,
+        callerId?: string,
+        retryCount = 7
+    ): Promise<void> {
+        t.Record({
+            pipeName: t.String,
+            logic: BotLogicT,
+            appId: t.String,
+            audience: t.String,
+            callerId: t.Optional(t.String),
+        }).check({ pipeName, logic, appId, audience, callerId });
+
+        // The named pipe is local and so there is no network authentication to perform: so we can create the result here.
+        const authenticateRequestResult: AuthenticateRequestResult = {
+            audience,
+            callerId,
+            claimsIdentity: appId ? this.createClaimsIdentity(appId) : new ClaimsIdentity([]),
+        };
+
+        // Creat request handler
+        const requestHandler = new StreamingRequestHandler(
+            authenticateRequestResult,
+            (authenticateRequestResult, activity) => this.processActivity(authenticateRequestResult, activity, logic)
+        );
+
+        // Create server
+        const server = new NamedPipeServer(pipeName, requestHandler);
+
+        // Attach server to request handler for outbound requests
+        requestHandler.server = server;
+
+        // Spin it up
+        await retry(() => server.start(), retryCount);
+    }
+
+    private async connect(req: Request, socket: INodeSocket, head: INodeBuffer, logic: BotLogic): Promise<void> {
+        // Grab the auth header from the inbound http request
+        const authHeader = t.String.check(req.headers.Authorization ?? req.headers.authorization ?? '');
+
+        // Grab the channelId which should be in the http headers
+        const channelIdHeader = t.Optional(t.String).check(req.headers.channelid);
+
+        // Authenticate inbound request
+        const authenticateRequestResult = await this.botFrameworkAuthentication.authenticateStreamingRequest(
+            authHeader,
+            channelIdHeader
+        );
+
+        // Creat request handler
+        const requestHandler = new StreamingRequestHandler(
+            authenticateRequestResult,
+            (authenticateRequestResult, activity) => this.processActivity(authenticateRequestResult, activity, logic)
+        );
+
+        // Create server
+        const server = new WebSocketServer(
+            await new NodeWebSocketFactory().createWebSocket(req, socket, head),
+            requestHandler
+        );
+
+        // Attach server to request handler
+        requestHandler.server = server;
+
+        // Spin it up
+        await server.start();
+    }
+}
+
+/**
+ * QUESTION(jpg): why doesn't this already exist? What is IStreamingActivityProcessor
```

Comment:
Some of these helpers we're direct or faithful ports from .NET as there are some streaming components that don't exist in JS. However, this felt like a happy medium. It seems to be working okay as well.